lib/web_ui/lib/src/engine/html/render_vertices.dart

```diff
@@ -124,60 +168,132 @@ class _WebGlRenderer implements _GlRenderer {
+        transform);
+
+    if (imageShader != null) {
+      /// To map from vertex position to texture coordinate in 0..1 range,
+      /// we setup scalar to be used in vertex shader.
+      _setupTextureScalar(gl, glProgram, 1.0 / imageShader._image.width.toDouble(),
+          1.0 / imageShader._image.height.toDouble());
+    }
 
     // Setup geometry.
+    //
+    // Create buffer for vertex coordinates.
     Object positionsBuffer = gl.createBuffer()!;
     assert(positionsBuffer != null); // ignore: unnecessary_null_comparison
+
+    Object? vao;
+    if (imageShader != null) {
+      if (isWebGl2) {
+        // Create a vertex array object.
+        vao = gl.createVertexArray();
+        // Set vertex array object as active one.
+        gl.bindVertexArray(vao!);
+      }
+    }
+    // Turn on position attribute.
+    gl.enableVertexAttribArray(positionAttributeLocation);
+    // Bind buffer as position buffer and transfer data.
     gl.bindArrayBuffer(positionsBuffer);
     gl.bufferData(positions, gl.kStaticDraw);
-    Object? positionLoc = gl.getAttributeLocation(glProgram.program, 'position');
+    // Setup data format for attribute.
     js_util.callMethod(
         gl.glContext, 'vertexAttribPointer', <dynamic>[
-          positionLoc, 2, gl.kFloat, false, 0, 0,
+      positionAttributeLocation, 2, gl.kFloat, false, 0, 0,
     ]);
-    gl.enableVertexAttribArray(0);
-
-    // Setup color buffer.
-    Object? colorsBuffer = gl.createBuffer();
-    gl.bindArrayBuffer(colorsBuffer);
 
     final int vertexCount = positions.length ~/ 2;
-
-    // Buffer kBGRA_8888.
-    if (vertices._colors == null) {
-      final ui.Color color = paint.color ?? ui.Color(0xFF000000);
-      Uint32List vertexColors = Uint32List(vertexCount);
-      for (int i = 0; i < vertexCount; i++) {
-        vertexColors[i] = color.value;
+    Object? texture;
+
+    if (imageShader == null) {
+      // Setup color buffer.
+      Object? colorsBuffer = gl.createBuffer();
+      gl.bindArrayBuffer(colorsBuffer);
+
+      // Buffer kBGRA_8888.
+      if (vertices._colors == null) {
+        final ui.Color color = paint.color ?? ui.Color(0xFF000000);
+        Uint32List vertexColors = Uint32List(vertexCount);
+        for (int i = 0; i < vertexCount; i++) {
+          vertexColors[i] = color.value;
+        }
+        gl.bufferData(vertexColors, gl.kStaticDraw);
+      } else {
+        gl.bufferData(vertices._colors, gl.kStaticDraw);
       }
-      gl.bufferData(vertexColors, gl.kStaticDraw);
+      Object colorLoc = gl.getAttributeLocation(glProgram.program, 'color');
+      js_util.callMethod(gl.glContext, 'vertexAttribPointer',
+          <dynamic>[colorLoc, 4, gl.kUnsignedByte, true, 0, 0]);
+      gl.enableVertexAttribArray(colorLoc);
     } else {
-      gl.bufferData(vertices._colors, gl.kStaticDraw);
+      // Copy image it to the texture.
+      texture = gl.createTexture();
+      // use texture unit 0 ??webgl2?
+      gl.activeTexture(gl.kTexture0 + 0);
```

Comment:
I don't understand the comment and `+ 0`?

Reply:
Thanks for catching that. Added a long comment there and removed +0.